File: lib/java/src/TransportFactories.java
package agnos;

import java.io.*;
import java.util.*;
import java.net.*;


public class TransportFactories
{
	public interface ITransportFactory 
	{
		void close() throws IOException;
		Transports.ITransport accept() throws IOException;
	}
	
	public static class SocketTransportFactory implements ITransportFactory 
	{
		protected int backlog;
		protected ServerSocket serverSocket;

		public SocketTransportFactory(int port) throws IOException {
			this(port, 10);
		}

		public SocketTransportFactory(int port, int backlog) throws IOException {
			this(InetAddress.getLocalHost(), port, backlog);
		}

		public SocketTransportFactory(String host, int port) throws IOException {
			this(host, port, 10);
		}

		public SocketTransportFactory(String host, int port, int backlog) throws IOException {
			this(InetAddress.getByName(host), port, backlog);
		}

		public SocketTransportFactory(InetAddress addr, int port, int backlog)
				throws IOException {
			serverSocket = new ServerSocket(port, backlog, addr);
			this.backlog = backlog;
		}

		public void close() throws IOException {
			serverSocket.close();
		}

		public Transports.ITransport accept() throws IOException {
			return new Transports.SocketTransport(serverSocket.accept());
		}
	}
	
}

File: lib/csharp/src/TransportFactories.cs
using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using Agnos.Transports;


namespace Agnos.TransportFactories
{
	public interface ITransportFactory
	{
		Transports.ITransport Accept();
		void Close();
	}

	public class SocketTransportFactory : ITransportFactory
	{
		internal TcpListener listener;
        public const int DefaultBacklog = 10;

		public SocketTransportFactory(int port) :
			this(IPAddress.Any, port)
		{
		}

		public SocketTransportFactory(int port, int backlog) :
			this(IPAddress.Any, port, backlog)
		{
		}

        protected static IPAddress GetIPv4AddressOf(String host)
        {
            foreach (IPAddress addr in Dns.GetHostEntry(host).AddressList) {
                if (addr.AddressFamily == AddressFamily.InterNetwork) {
                    return addr;
                }
            }
            return null;
        }

		public SocketTransportFactory(String host, int port) :
            this(GetIPv4AddressOf(host), port)
		{
		}

		public SocketTransportFactory(String host, int port, int backlog) :
            this(GetIPv4AddressOf(host), port, backlog)
		{
		}

        public SocketTransportFactory(IPAddress addr, int port) :
            this(addr, port, DefaultBacklog)
        {
        }

        public SocketTransportFactory(IPAddress addr, int port, int backlog)
        {
            listener = new TcpListener(addr, port);
            listener.Start(backlog);
        }

		public ITransport Accept()
		{
			return new SocketTransport(listener.AcceptSocket());
		}
		
		public void Close()
		{
			listener.Stop();
		}
	}
	
}